### DISCUSSION TITLE
Install on a remote Virtual Machine

### DISCUSSION THREAD

**Discussion by silviacesari95** [giongto35/cloud-game#284] [Q&A] [answered]
I am trying to install the application on an aws virtual machine through docker-compose. I was able to let it run but when I connect to the port 8000 on the vm public IP,  it starts the interface and the webrtc connection seems to be established, but doesn't appear the list of games and it seems all stuck.

`cloud-game-local | I0223 16:44:39.505355       9 webrtc.go:195] ICE Connection State has changed: connected
cloud-game-local | I0223 16:44:39.506173       9 webrtc.go:199] ConnectionStateConnected
cloud-game-local | I0223 16:44:39.506184       9 webrtc.go:334] Start streaming
cloud-game-local | I0223 16:44:39.702214       9 webrtc.go:169] Data channel 'game-input'-'824636433032' open.
`
I opened the related udp and tcp ports from aws, but there is something missing, I'm stuck here since all day, I don't have anymore ideas, please give me any suggestions. I also tried replacing all localhost with vm public IP.

**Comment by sergystepanov** [comment, marked as answer]
Hard to tell, would be great to see a complete log (docker-compose logs).
The last time I checked Docker Compose on a DO instance, it was running with only one change in the Coordinator address.
For WebRTC you have to have all your cloud machine ports open.

**Reply by silviacesari95** [reply]
this is my docker-compse:
version: '3'
services:

  cloud-game:
    build: .
    image: cloud-game-local
    container_name: cloud-game-local
    privileged: true
    environment:
      - DISPLAY=:99
      - MESA_GL_VERSION_OVERRIDE=3.3
    network_mode: host
    command: >
      bash -c "Xvfb :99 & coordinator --v=5 & worker --coordinatorhost=public-ip:8000"
    volumes:
      # keep cores persistent in the cloud-game_cores volume
      - cores:/usr/local/share/cloud-game/assets/cores
      - games:/usr/local/share/cloud-game/assets/games

volumes:
  cores:
  games:





[ec2-user@ip-10-0-0-6 cloud-game]$ docker-compose up
Building with native build. Learn about native build in Compose here: https://docs.docker.com/go/compose-native-build/
Creating cloud-game-local ... done
Attaching to cloud-game-local
cloud-game-local | I0225 09:07:01.273837       7 main.go:30] Initializing coordinator server
cloud-game-local | I0225 09:07:01.273904       7 main.go:31] Coordinator configs {{http://localhost:8000   {assets/games [] [neogeo pgm] true false} {6601 /coordinator false false} {8000 443  }} {1 {false 320 240} {user}/.cr/save {{{assets/cores assets/cores} {true {buildbot https://buildbot.libretro.com/nightly zip} {github https://github.com/sergystepanov/libretro-spiegel/blob/main zip}} map[gba:{mgba_libretro  [gba gbc] 240 160 0 false false false false} n64:{mupen64plus_next_libretro mupen64plus_next_libretro.cfg [n64 v64 z64] 320 240 0 true true false false} nes:{nestopia_libretro  [nes] 256 240 0 false false false false} pcsx:{pcsx_rearmed_libretro pcsx_rearmed_libretro.cfg [cue] 350 240 0 false false false false} snes:{snes9x_libretro  [smc sfc swc fig bs] 256 224 0 false false true false}]}}} dev {[{stun:stun.l.google.com:19302  }] {0 0} 10.5.0.6}}
cloud-game-local | I0225 09:07:01.277155       7 game_library.go:141] [lib] scan... started
cloud-game-local | I0225 09:07:01.277388       7 game_library.go:262] 
cloud-game-local | --------------------------------------------
cloud-game-local | --- The Library of ROMs                  ---
cloud-game-local | --------------------------------------------
cloud-game-local | --------------------------------------------
cloud-game-local | --- ROMs: 000                   38.982µs ---
cloud-game-local | --------------------------------------------
cloud-game-local | I0225 09:07:01.277403       7 game_library.go:184] [lib] scan... completed
cloud-game-local | I0225 09:07:01.277416       7 coordinator.go:113] Initializing Coordinator Server
cloud-game-local | I0225 09:07:01.277788       7 coordinator.go:47] Starting monitoring server for coordinator
cloud-game-local | I0225 09:07:01.278441       8 worker.go:52] [worker] active modules: [monitoring::/worker:6601]
cloud-game-local | I0225 09:07:01.278662       8 handlers.go:213] Set storage: /root/.cr/save
cloud-game-local | I0225 09:07:01.279974       8 monitoring.go:41] [worker] Starting monitoring server at :6601
cloud-game-local | I0225 09:07:01.280076       8 monitoring.go:63] [worker] Prometheus metric is enabled at :6601/worker/metrics
cloud-game-local | I0225 09:07:01.280567       8 storage.go:26] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
cloud-game-local | I0225 09:07:01.280674       8 handlers.go:89] Starting Libretro cores sync...
cloud-game-local | I0225 09:07:01.281041       8 handlers.go:106] Worker connecting to coordinator: ws://148.252.131.35:8000/wso?zone=
cloud-game-local | I0225 09:07:01.281702       7 handlers.go:82] Coordinator: A worker is connecting...
cloud-game-local | I0225 09:07:01.281897       8 handlers.go:66] [worker] connected to: 148.252.131.35:8000
cloud-game-local | I0225 09:07:01.282126       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Generated worker ID
cloud-game-local | I0225 09:07:01.282143       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Address:148.252.131.35
cloud-game-local | I0225 09:07:01.282164       7 worker.go:34] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Is public: true zone: 
cloud-game-local | I0225 09:07:01.282170       7 worker.go:34] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Set ping server address: http://148.252.131.35:9000/echo
cloud-game-local | I0225 09:07:01.282671       8 internalhandlers.go:18] [worker] new id: 535dc79d-b569-4353-bb7e-64d1636bfdc5
cloud-game-local | I0225 09:07:11.316685       7 handlers.go:151] Coordinator: A user is connecting...
cloud-game-local | I0225 09:07:11.316770       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Generated worker ID
cloud-game-local | I0225 09:07:11.316827       7 browser.go:29] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Get Room  Zone  From URL /ws?room_id=&zone=
cloud-game-local | I0225 09:07:11.316844       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Send sync[http://148.252.131.35:9000/echo]http://148.252.131.35:9000/echo
cloud-game-local | I0225 09:07:11.420452       8 server.go:33] &{0xc000358640 0xc000149200 {} 0x4ef360 false false false false 0 {0 0} 0xc0003568c0 {0xc0004a41c0 map[] false false} map[Access-Control-Allow-Origin:[*]] true 0 -1 0 false false [] 0 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [0 0 0 0 0 0 0 0 0 0] [0 0 0] 0xc00016a700 0} echo
cloud-game-local | I0225 09:07:11.486087       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Latency mapmap[0xc000076720:111]
cloud-game-local | I0225 09:07:11.537044       7 browser.go:29] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received init_webrtc request -> relay to worker: 535dc79d-b569-4353-bb7e-64d1636bfdc5
cloud-game-local | I0225 09:07:11.537328       8 internalhandlers.go:44] Received a request to createOffer from browser via coordinator
cloud-game-local | I0225 09:07:11.537520       8 webrtc.go:120] === StartClient ===
cloud-game-local | I0225 09:07:11.545830       8 webrtc.go:141] Add video track
cloud-game-local | I0225 09:07:11.547224       8 webrtc.go:231] Created Offer
cloud-game-local | I0225 09:07:11.548746       8 internalhandlers.go:72] Start peerconnection fa376546-7fdc-4320-8d3f-affc7e578a01
cloud-game-local | I0225 09:07:11.549301       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received SDP from worker -> sending back to browser
cloud-game-local | I0225 09:07:11.549764       8 webrtc.go:192] OnIceCandidate: candidate:2216991393 1 udp 2130706431 10.5.0.6 39651 typ host
cloud-game-local | I0225 09:07:11.549904       8 webrtc.go:192] OnIceCandidate: candidate:2216991393 1 udp 2130706431 10.5.0.6 55614 typ host
cloud-game-local | I0225 09:07:11.550002       8 webrtc.go:192] OnIceCandidate: candidate:2216991393 1 udp 2130706431 10.5.0.6 39284 typ host
cloud-game-local | I0225 09:07:11.550461       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Received IceCandidate from worker -> relay to browser
cloud-game-local | I0225 09:07:11.550872       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Received IceCandidate from worker -> relay to browser
cloud-game-local | I0225 09:07:11.551181       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Received IceCandidate from worker -> relay to browser
cloud-game-local | I0225 09:07:11.563079       8 webrtc.go:192] OnIceCandidate: candidate:573745292 1 udp 1694498815 148.252.131.35 41790 typ srflx raddr 0.0.0.0 rport 41790
cloud-game-local | I0225 09:07:11.563760       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Received IceCandidate from worker -> relay to browser
cloud-game-local | I0225 09:07:11.563906       7 worker.go:38] Worker 535dc79d-b569-4353-bb7e-64d1636bfdc5] Received IceCandidate from worker -> relay to browser
cloud-game-local | I0225 09:07:11.641561       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received browser answered SDP -> relay to worker
cloud-game-local | I0225 09:07:11.642338       8 internalhandlers.go:88] Received answer SDP from browser
cloud-game-local | I0225 09:07:11.643346       8 webrtc.go:264] Set Remote Description
cloud-game-local | I0225 09:07:11.643445       8 webrtc.go:176] ICE Connection State has changed: checking
cloud-game-local | I0225 09:07:11.659433       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received IceCandidate from browser -> relay to worker
cloud-game-local | I0225 09:07:11.659959       8 internalhandlers.go:106] Received remote Ice Candidate from browser
cloud-game-local | I0225 09:07:11.660407       8 webrtc.go:275] Decoded Ice: candidate:1418740086 1 udp 2113937151 42a6eabb-9343-4ebf-b500-ec2f5eaf5d37.local 47210 typ host generation 0 ufrag J4P1 network-cost 999
cloud-game-local | I0225 09:07:11.660499       8 webrtc.go:283] Add Ice Candidate: candidate:1418740086 1 udp 2113937151 42a6eabb-9343-4ebf-b500-ec2f5eaf5d37.local 47210 typ host generation 0 ufrag J4P1 network-cost 999
cloud-game-local | I0225 09:07:11.680482       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received IceCandidate from browser -> relay to worker
cloud-game-local | I0225 09:07:11.680964       8 internalhandlers.go:106] Received remote Ice Candidate from browser
cloud-game-local | I0225 09:07:11.681023       8 webrtc.go:275] Decoded Ice: candidate:842163049 1 udp 1677729535 151.81.99.127 47210 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag J4P1 network-cost 999
cloud-game-local | I0225 09:07:11.681055       8 webrtc.go:283] Add Ice Candidate: candidate:842163049 1 udp 1677729535 151.81.99.127 47210 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag J4P1 network-cost 999
cloud-game-local | I0225 09:07:12.326327       8 webrtc.go:176] ICE Connection State has changed: connected
cloud-game-local | I0225 09:07:12.326408       8 webrtc.go:180] ConnectionStateConnected
cloud-game-local | I0225 09:07:12.326427       8 webrtc.go:315] Start streaming
cloud-game-local | I0225 09:07:12.512242       8 webrtc.go:160] Data channel 'game-input'-'824633952876' open.
cloud-game-local | I0225 09:07:15.340538       7 browser.go:33] Browser fa376546-7fdc-4320-8d3f-affc7e578a01] Received start request from a browser -> relay to worker

**Reply by sergystepanov** [reply]
I see. 👀 It seems that for some reason it can't find any games inside the container (`/usr/local/share/cloud-game/assets/games`).
```
--- ROMs: 000 38.982µs ---
```
This is happening because you've changed the games folder mapping. Our Docker image doesn't copy any games during the build and expects that you map a folder with games on your host machine. Something like that:
```
CLOUD_GAME_GAMES_PATH=$(PWD)/assets/games docker-compose up --build
```
So, either add the copy command for games into Dokcerfile (look there at the end) or just use original mapping with the env variable.
Also, you can just leave it as is and copy your games inside your current image since it's persistent.

I'll look into more user-friendly logging and error messaging sometime later.

**Reply by silviacesari95** [reply]
Great! It works, thank you very much! :)
